### PR TITLE
[E2E] Cypress Replay.io Integration

### PR DESCRIPTION
**Pull request by nemanjaglumac**
https://www.replay.io/

This PR integrates Replay.io into our E2E automated tests workflow.
After different trials and setups, I've determined in cooperation with replay's core dev team it's best to run these recordings on a schedule in the beginning as an experiment.
As shown by [this latest run](https://github.com/metabase/metabase/actions/runs/5084940536), this should in no way interfere with our ability to run E2E tests as we did before - on pull request and on merge to master or release branches.

This is the E2E run on PR:
```
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        12.8.1                                                                         │
  │ Browser:        Chrome 113 (headless)                                                          │
  │ Node Version:   v16.10.0 (/opt/hostedtoolcache/node/16.10.0/x64/bin/node)                      │
  │ Specs:          2 found (default.cy.snap.js, qa-db.cy.snap.js)                                 │
  │ Searched:       e2e/snapshot-creators/**/*.cy.snap.js                                          │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

And this is how it looks like when we enable Replay (and how it will look like on schedule):
```
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        12.8.1                                                                         │
  │ Browser:        Replay 91 (headless)                                                           │
  │ Node Version:   v16.10.0 (/opt/hostedtoolcache/node/16.10.0/x64/bin/node)                      │
  │ Specs:          2 found (default.cy.snap.js, qa-db.cy.snap.js)                                 │
  │ Searched:       e2e/snapshot-creators/**/*.cy.snap.js                                          │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Replay runs are skipped during "regular" checks (pr/merge)
image

### CODE REVIEW

**Comment by ryanjduffy** [comment]
👋 from Replay! I'm not sure what's happening here but it appears that the replay browsers aren't being installed. Usually that happens automatically when the `@replayio/cypress` package is installed but there may be a bug in that logic.

In the mean time, you can manually trigger an install using `npx @replayio/cypress install` in your cypress setup action. I think that'll resolve the `writeFileSync` issue because it's trying to write to the replay "home" directory which is usually created by the browser install process. I'll open a ticket internally to handle this scenario better and report a more useful error.

File: .github/workflows/e2e-tests.yml [hunk 1]
```diff
@@ -183,6 +185,14 @@ jobs:
         env:
           TERM: xterm
 
+      - name: Upload Replay.io recordings
+        if: failure()
+        uses: replayio/action-upload@v0.4.7
+        with:
+          api-key: ${{ secrets.RECORD_REPLAY_API_KEY }}
+          filter: ${{ 'function($v) { $v.metadata.test.result = "failed" }' }}
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:193]:
```suggestion
```

Comment (jaril) [reply .github/workflows/e2e-tests.yml:193]: Let's remove this filter for now. It's helpful to get all of the replays just so we can spot check them.

File: .github/workflows/e2e-tests.yml [hunk 2]
```diff
@@ -183,6 +185,14 @@ jobs:
         env:
           TERM: xterm
 
+      - name: Upload Replay.io recordings
+        if: failure()
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:189]:
```suggestion
```

Comment (jaril) [reply .github/workflows/e2e-tests.yml:189]: Let's make this into if: always(), just so it doesn't bail when the previous step fails.

File: .github/workflows/e2e-tests.yml [hunk 3]
```diff
@@ -183,6 +185,13 @@ jobs:
         env:
           TERM: xterm
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:186]:
```suggestion
          TERM: xterm
          DEBUG: replay:cypress:plugin:events,replay:test-utils:reporter
```

Comment (jaril) [reply .github/workflows/e2e-tests.yml:186]: @nemanjaglumac mind adding this debug envvar so we can take a look at the logs?

We seem to be missing test metadata, so we're suspicious that the plugin isn't working as it should. This should confirm some of those suspicions.

File: .github/workflows/e2e-tests.yml [hunk 4]
```diff
@@ -184,7 +184,8 @@ jobs:
           --folder ${{ matrix.folder }}
         env:
           TERM: xterm
-          DEBUG: replay:cypress:plugin:events,replay:test-utils:reporter
+          DEBUG: replay:cypress:plugin:events,replay:test-utils:reporter,cypress:launcher*
+          RECORD_REPLAY_VERBOSE: 1
```
Comment (ryanjduffy) [inline .github/workflows/e2e-tests.yml:188]: I've discovered that DeploySentinel does not pass on environment variables set by our plugin on to the browser. I'll follow up with them on that but, in the mean time, you can set the following env variable to get started.

```suggestion
          RECORD_REPLAY_METADATA_FILE: /tmp/replay-metadata.json
```

**Comment by jaril** [comment]
@nemanjaglumac pushed a [separate PR](https://github.com/metabase/metabase/pull/30670) to consolidate the tests. Right now they're being displayed as multiple test runs in the Replay UI. This changes makes it so that there's only one test run entry per github action run.

**Comment by jaril** [comment]
@nemanjaglumac mind adding this diff in and re-running the tests??

```
diff --git a/.github/workflows/e2e-tests.yml b/.github/workflows/e2e-tests.yml
index 195efc2668..fa3f7f5d52 100644
--- a/.github/workflows/e2e-tests.yml
+++ b/.github/workflows/e2e-tests.yml
@@ -182,6 +182,9 @@ jobs:
           --spec './e2e/test/scenarios/**/*.cy.spec.js'
         env:
           TERM: xterm
+          RECORD_REPLAY_METADATA_FILE: /tmp/replay-metadata.json
+          RECORD_REPLAY_METADATA_TEST_RUN_ID: ${{ needs.test-run-id.outputs.testRunId }}
+          RECORD_REPLAY_DISABLE_ASSERTS: 1

       - name: Run EE Cypress tests on ${{ matrix.folder }}
         if: matrix.edition == 'ee'
```

**Comment by jaril** [comment]
One more time saving we can try, can you add this line as well?

```
diff --git a/.github/workflows/e2e-tests.yml b/.github/workflows/e2e-tests.yml
index fa3f7f5d52..54f2168799 100644
--- a/.github/workflows/e2e-tests.yml
+++ b/.github/workflows/e2e-tests.yml
@@ -185,6 +185,7 @@ jobs:
           RECORD_REPLAY_METADATA_FILE: /tmp/replay-metadata.json
           RECORD_REPLAY_METADATA_TEST_RUN_ID: ${{ needs.test-run-id.outputs.testRunId }}
           RECORD_REPLAY_DISABLE_ASSERTS: 1
+          RECORD_REPLAY_DISABLE_SOURCEMAP_COLLECTION: 1

       - name: Run EE Cypress tests on ${{ matrix.folder }}
         if: matrix.edition == 'ee'
(END)
```

File: .github/workflows/e2e-tests.yml [hunk 5]
```diff
@@ -173,6 +182,9 @@ jobs:
           --spec './e2e/test/scenarios/**/*.cy.spec.js'
         env:
           TERM: xterm
+          RECORD_REPLAY_METADATA_FILE: /tmp/replay-metadata.json
+          RECORD_REPLAY_METADATA_TEST_RUN_ID: ${{ needs.test-run-id.outputs.testRunId }}
+          RECORD_REPLAY_DISABLE_ASSERTS: 1
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:187]:
```suggestion
          RECORD_REPLAY_DISABLE_ASSERTS: 1
          RECORD_REPLAY_DISABLE_SOURCEMAP_COLLECTION: 1
```

File: .github/workflows/e2e-tests.yml [hunk 6]
```diff
@@ -182,6 +194,16 @@ jobs:
           --folder ${{ matrix.folder }}
         env:
           TERM: xterm
+          RECORD_REPLAY_METADATA_FILE: /tmp/replay-metadata.json
+          RECORD_REPLAY_METADATA_TEST_RUN_ID: ${{ needs.test-run-id.outputs.testRunId }}
+          RECORD_REPLAY_DISABLE_ASSERTS: 1
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:199]:
```suggestion
          RECORD_REPLAY_DISABLE_ASSERTS: 1
          RECORD_REPLAY_DISABLE_SOURCEMAP_COLLECTION: 1
```

File: .github/workflows/e2e-tests.yml [hunk 7]
```diff
@@ -171,17 +180,34 @@ jobs:
           yarn run test-cypress-run \
           --env grepTags=@OSS \
           --spec './e2e/test/scenarios/**/*.cy.spec.js'
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:182]:
```suggestion
          --spec './e2e/test/scenarios/**/*.cy.spec.js' \
```

File: .github/workflows/e2e-tests.yml [hunk 8]
```diff
@@ -170,18 +179,37 @@ jobs:
...
-          --spec './e2e/test/scenarios/**/*.cy.spec.js'
+          --spec './e2e/test/scenarios/**/*.cy.spec.js' \
+          --browser "replay-chromium"
         env:
           TERM: xterm
+          REPLAYIO_ENABLED: 1
+          RECORD_REPLAY_METADATA_FILE: /tmp/replay-metadata.json
+          RECORD_REPLAY_METADATA_TEST_RUN_ID: ${{ needs.test-run-id.outputs.testRunId }}
+          RECORD_REPLAY_DISABLE_ASSERTS: 1
+          RECORD_REPLAY_DISABLE_SOURCEMAP_COLLECTION: 1
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:190]:
```suggestion
```

File: .github/workflows/e2e-tests.yml [hunk 9]
```diff
@@ -170,18 +179,37 @@ jobs:
...
-          --folder ${{ matrix.folder }}
+          --folder ${{ matrix.folder }} \
+          --browser "replay-chromium"
         env:
           TERM: xterm
+          REPLAYIO_ENABLED: 1
+          RECORD_REPLAY_METADATA_FILE: /tmp/replay-metadata.json
+          RECORD_REPLAY_METADATA_TEST_RUN_ID: ${{ needs.test-run-id.outputs.testRunId }}
+          RECORD_REPLAY_DISABLE_ASSERTS: 1
+          RECORD_REPLAY_DISABLE_SOURCEMAP_COLLECTION: 1
```
Comment (jaril) [inline .github/workflows/e2e-tests.yml:205]:
```suggestion
```

**Review by jaril — COMMENTED** [review]
let's take these out since the overhead from sourcemap collection should be a non-concern by EOW

**Review by npfitz — APPROVED** [review]

**Pull request merged**